### PR TITLE
Add uefi support on ubuntu

### LINKED ISSUE #1743: add shim support for ubuntu (>=20.04)
## Problem description
i try to build a ubuntu 20.04 uefi VM image and the build fails with `KiwiBootLoaderGrubSecureBootError: Signed grub2 efi loader not found`.
If i compare that with the example build from https://build.opensuse.org/package/live_build_log/Virtualization:Appliances:Images:Testing_x86:ubuntu/test-image-live-disk:Virtual/images/x86_64 the log differs quite a lot.

```
kiwi --debug --color-output --shared-cache-dir=./kiwi-cache --profile=vm_rescue_uefi system build --description=descriptions/ubuntu-20.04-amd64-vm_rescue_uefi --target-dir=out/ubuntu-20.04-amd64-vm_rescue_uefi

...
[ DEBUG   ]: 20:20:20 | EXEC: [blkid /dev/mapper/loop0p3 -s UUID -o value]
[ DEBUG   ]: 20:20:20 | EXEC: [blkid /dev/mapper/loop0p3 -s UUID -o value]
[ INFO    ]: 20:20:20 | Creating grub2 bootloader images
[ DEBUG   ]: 20:20:20 | EXEC: [mkdir -p /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/boot/efi/EFI/BOOT]
[ DEBUG   ]: 20:20:20 | EXEC: [mkdir -p /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/boot/grub/fonts]
[ DEBUG   ]: 20:20:20 | EXEC: [cp /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/usr/share/grub/unicode.pf2 /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/boot/grub/fonts]
[ DEBUG   ]: 20:20:20 | EXEC: [mkdir -p /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/boot/grub/themes]
[ DEBUG   ]: 20:20:20 | EXEC: [rsync -a --exclude /*.module /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/usr/lib/grub/i386-pc/ /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/boot/grub/i386-pc]
[ DEBUG   ]: 20:20:20 | EXEC: [rsync -a --exclude /*.module /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/usr/lib/grub/x86_64-efi/ /buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/boot/grub/x86_64-efi]
[ DEBUG   ]: 20:20:20 | "shim-install": in paths "/buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/usr/local/sbin:/buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/usr/local/bin:/buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/usr/sbin:/buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/usr/bin:/buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/sbin:/buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/bin:/buildpath/kiwi/out/ubuntu-20.04-amd64-vm_rescue_uefi/build/image-root/snap/bin" exists: "False" mode match: not checked
[ WARNING ]: 20:20:20 | --> Running fallback setup for shim secure boot efi image
[ ERROR   ]: 20:20:20 | KiwiBootLoaderGrubSecureBootError: Signed grub2 efi loader not found
[ INFO    ]: 20:20:20 | Cleaning up FileSystemExt4 instance
[ INFO    ]: 20:20:20 | Cleaning up FileSystemFat16 instance
...
```
## Expected behaviour
the image is built as in the link above

## Steps to reproduce the behaviour

```xml
<?xml version="1.0" encoding="utf-8"?>
<image schemaversion="7.3" name="ubuntu-20.04-amd64-CURRENT_BUILD_IMAGETYPE">
  <description type="system">
    <author>B1 Systems GmbH</author>
    <contact>DL CCloud Managed Service (External) DL_5B3B97BF5F99B7139C00005D@sap.com</contact>
    <specification>oem disk build for Ubuntu 20.04</specification>
  </description>
  <profiles>
    <profile name="vmware" description="VM image for vmware" import="true"/>
    <profile name="vm_rescue_uefi" description="VM rescue image for UEFI" import="true"/>
  </profiles>
  <preferences>
    <version>CURRENT_BUILD_VERSION</version>
    <packagemanager>apt-get</packagemanager>
    <bootsplash-theme/>
    <bootloader-theme/>
    <locale>en_US</locale>
    <keytable>us</keytable>
    <timezone>UTC</timezone>
    <rpm-check-signatures>false</rpm-check-signatures>
  </preferences>
  <preferences profiles="vmware">
    <type image="oem" filesystem="ext4" format="vmdk" kernelcmdline="earlyprintk=vga product_name='OpenStack Nova' consoleblank=0" firmware="bios">
      <machine guestOS="ubuntu64guest" arch="x86_64" ovftype="vmware">
        <vmdisk diskmode="streamOptimized"/>
      </machine>
      <bootloader name="grub2" console="console"/>
      <oemconfig>
        <oem-resize>false</oem-resize>
      </oemconfig>
    </type>
  </preferences>

  <preferences profiles="vm_rescue_uefi">
    <type image="oem" filesystem="ext4" format="vmdk" kernelcmdline="earlyprintk=efi console=tty0 console=ttyS0,115200n8 product_name='OpenStack Nova'" firmware="uefi">
      <machine guestOS="ubuntu64guest" arch="x86_64" ovftype="vmware">
        <vmdisk diskmode="streamOptimized"/>
      </machine>
      <bootloader name="grub2" console="console"/>
      <oemconfig>
        <oem-resize>false</oem-resize>
      </oemconfig>
    </type>
  </preferences>

  <users profiles="vmware">
    <user home="/root" name="root" groups="root" password="" pwdformat="plain" shell="/bin/bash"/>
    <user home="/home/ccloud" name="ccloud" groups="ccloud,sudo" password="" pwdformat="plain" shell="/bin/bash"/>
  </users>

  <users profiles="vm_rescue_uefi">
    <user home="/root" name="root" groups="root" password="$6$hash" pwdformat="encrypted" shell="/bin/bash"/>
    <user home="/home/ccloud" name="ccloud" groups="ccloud,sudo" password="$6$hash" pwdformat="plain" shell="/bin/bash"/>
  </users>
  <repository type="apt-deb" alias="Focal-security" distribution="focal-security" components="main multiverse restricted universe" imageinclude="false" repository_gpgcheck="false">
    <source path="...">
  </repository>

  <packages type="image" patternType="plusRecommended">
    <package name="apport"/>
    <package name="at"/>
    <package name="bcache-tools"/>
    <package name="btrfs-progs"/>
    <package name="byobu"/>
    <package name="cloud-init"/>
    <package name="curl"/>
    <package name="dirmngr"/>
    <package name="dracut"/>
    <package name="ethtool"/>
    <package name="fonts-ubuntu-console"/>
    <package name="fwupd"/>
    <package name="git"/>
    <package name="gnupg"/>
    <package name="htop"/>
    <package name="iputils-arping"/>
    <package name="landscape-common"/>
    <package name="language-pack-en"/>
    <package name="libnss-myhostname"/>
    <package name="libnss-resolve"/>
    <package name="linux-image-generic"/>
    <package name="lvm2"/>
    <package name="lxd-agent-loader"/>
    <package name="manpages-posix-dev"/>
    <package name="mdadm"/>
    <package name="mlocate"/>
    <package name="open-iscsi"/>
    <package name="open-vm-tools"/>
    <package name="openssh-server"/>
    <package name="patch"/>
    <package name="policykit-1"/>
    <package name="pollinate"/>
    <package name="screen"/>
    <package name="snapd"/>
    <package name="software-properties-common"/>
    <package name="sosreport"/>
    <package name="tmux"/>
    <package name="ubuntu-minimal"/>
    <package name="ubuntu-standard"/>
    <package name="update-notifier-common"/>
    <package name="vim"/>
    <package name="xfsprogs"/>
    <package name="xz-utils"/>
  </packages>

  <packages type="image" profiles="vm_rescue_uefi">
    <package name="grub-efi-amd64"/>
    <package name="grub-efi-amd64-bin"/>
    <package name="grub-efi-amd64-signed"/>
    <package name="grub-pc-bin"/>
    <package name="grub2-common"/>
  </packages>
  <packages type="bootstrap">
    <package name="apt-utils"/>
    <package name="apt-transport-https"/>
  </packages>
</image>
```

## OS and Software information
+ KIWI version: v9.23.16
+ Operating system: ubuntu 20.04
+ OBS version: none

### PR DESCRIPTION
**Pull request by schaefi**
Added support for UEFI on Ubuntu
    
The Ubuntu folks have a different system to support EFI secure boot. In order to make use of it kiwi needs some adaptions done in this pull request. This Fixes #1743

### CODE REVIEW

**Comment by schaefi** [comment]
The integration test building a secure boot image is here:

* https://build.opensuse.org/package/show/Virtualization:Appliances:Images:Testing_x86:ubuntu/test-image-live-disk

The ```Virtual``` Profile

**Comment by schaefi** [comment]
@davidcassany I'm not super happy with the pull request but I also don't see another solution. The part that Ubuntu provides the signed grub loader with the extension ```.signed``` which is something an EFI firmware does not understand can be easily handled by kiwi. The problem that the packaging on Ubuntu does not provide a vendor directory but requires one to find the config file is more of an issue. I solved this part by directly looking into the signed grub binary to see the hardcoded lookup path.

That part of the PR I don't like but I also don't know how we should fix this smarter.

Any feedback is welcome :)

Thanks

**Comment by schaefi** [comment]
I have tested:

```
kvm -bios /usr/share/qemu/ovmf-x86_64-ms.bin -hda kiwi-test-image-live-disk.x86_64-1.16.4-Virtual-Build11.2.qcow2 -serial stdio
```

And it works.

Please note: On Ubuntu you have to press ```Esc``` to display the menu. If you don't do that the system boots after the timeout but you don't see anything from grub. The integration test from above is also configured to use the serial terminal (bootloader_console="serial") This means when running the above command you need to set the focus to the shell window with the kvm call and hit ```Esc``` after the EFI stage was passed if you want to see the menu :)

File: kiwi/defaults.py [hunk 1]
```diff
@@ -860,7 +892,8 @@ def get_efi_vendor_directory(efi_path):
             'EFI/fedora',
             'EFI/redhat',
             'EFI/centos',
-            'EFI/opensuse'
+            'EFI/opensuse',
+            'EFI/ubuntu'
```
Comment (Conan-Kudo) [inline kiwi/defaults.py:896]: Debian uses this system too.

**Review by Conan-Kudo — CHANGES_REQUESTED** [review]

File: kiwi/defaults.py (hunk 1 above)
Comment (schaefi) [reply kiwi/defaults.py:896]: so it's ```EFI/debian``` or different ?

Comment (Conan-Kudo) [reply kiwi/defaults.py:896]: `EFI/debian`

Comment (schaefi) [reply kiwi/defaults.py]: done

**Review by davidcassany — APPROVED** [review]
Yes, this is convoluted. It took me a while to follow the logic of filename vs binaryname and all that.

Beyond that concern about having to actually parse the binary I think this is all good. Convoluted but not intrusive and I also do not see any easy and clean way of handling this bootloader discrepancies among distros.

Did you give it a test for Debian? It should be pretty easy to test by slightly adjusting the integration test.

**Comment by schaefi** [comment]
> Did you give it a test for Debian?

not yet, I'll adapt the Debian integration test as well... then we'll see if it works

**Comment by schaefi** [comment]
Integration test for Debian updated.

* https://build.opensuse.org/package/show/Virtualization:Appliances:Images:Testing_x86:debian/test-image-live-disk

Waiting for the build to finish

**Comment by schaefi** [comment]
There is no shim on Debian...

```
<buildinfo project="Virtualization:Appliances:Images:Testing_x86:debian" repository="images" package="test-image-live-disk:Virtual" downloadurl="https://download.opensuse.org/repositories">
  <arch>x86_64</arch>
  <error>unresolvable: nothing provides shim</error>
```

**Comment by schaefi** [comment]
We use Debian buster for the test

**Comment by Conan-Kudo** [comment]
`shim-signed` is present there: https://packages.debian.org/buster/shim-signed

**Comment by Conan-Kudo** [comment]
There is no `shim` package in Debian.

**Comment by davidcassany** [comment]
> `shim-signed` is present there: https://packages.debian.org/buster/shim-signed

Oh was about to say the same, @Conan-Kudo  you were faster than me :)

File: kiwi/defaults.py [hunk 2]
```diff
@@ -690,7 +699,8 @@ def get_shim_loader(root_path):
         shim_file_patterns = [
             '/usr/share/efi/*/shim.efi',
             '/usr/lib64/efi/shim.efi',
-            '/boot/efi/EFI/*/shim.efi'
+            '/boot/efi/EFI/*/shim.efi',
+            '/usr/lib/shim/shimx64.efi'
```
Comment (Conan-Kudo) [inline kiwi/defaults.py:703]: This should be `/usr/lib/shim/shim*.efi` as the name differs on the architecture.

**Comment by schaefi** [comment]
It's building now... I have my doubts that it will work. God I would wish there would a be a clear design about secure boot and a standard implementation for all distros. Actually nobody is following any reference implementation even though there is a standard paper for EFI and UEFI (Secure Boot) but everyone cooks its own soup on top of the Microsoft signed key ...

**Comment by Conan-Kudo** [comment]
We should also probably fix `/boot/efi/EFI/*/shim.efi` to be `/boot/efi/EFI/*/shim*.efi` in a followup PR, since distributions are dropping `shim.efi` for `shim<efiarch>.efi` name binaries for multiarch EFI support.

For example, on Fedora, both 32-bit and 64-bit UEFI shim binaries are installed for x86_64 to support 32-bit UEFI booting 64-bit OS.

File: kiwi/defaults.py (hunk 2 above)
Comment (schaefi) [reply kiwi/defaults.py]: done

**Comment by schaefi** [comment]
> We should also probably fix `/boot/efi/EFI/*/shim.efi` to be `/boot/efi/EFI/*/shim*.efi` in a followup PR, since distributions are dropping `shim.efi` for `shim<efiarch>.efi` name binaries for multiarch EFI support.
> 
> For example, on Fedora, both 32-bit and 64-bit UEFI shim binaries are installed for x86_64 to support 32-bit UEFI booting 64-bit OS.

yes good point, in an extra request please. Could you open an issue for this one please ? Thanks

**Comment by schaefi** [comment]
And the Debian build failed

```
[  382s] [ ERROR   ]: 13:04:11 | KiwiBootLoaderGrubSecureBootError: Signed grub2 efi loader not found
```

Actually this is expected. You said there is no shim package on Debian, so what package provides shim ?
The shim-signed package provides the grub parts with the shim signed loader. But the shim itself needs to be provided somehow.

**Comment by Conan-Kudo** [comment]
You need `grub-efi-amd64-signed` too.

**Comment by Conan-Kudo** [comment]
> > We should also probably fix `/boot/efi/EFI/*/shim.efi` to be `/boot/efi/EFI/*/shim*.efi` in a followup PR, since distributions are dropping `shim.efi` for `shim<efiarch>.efi` name binaries for multiarch EFI support.
> > For example, on Fedora, both 32-bit and 64-bit UEFI shim binaries are installed for x86_64 to support 32-bit UEFI booting 64-bit OS.
> 
> yes good point, in an extra request please. Could you open an issue for this one please ? Thanks

Done: https://github.com/OSInside/kiwi/issues/1806

**Comment by schaefi** [comment]
> You need `grub-efi-amd64-signed` too.

Thanks much that worked. The Debian image is now also working in UEFI mode :smiley:

**Comment by schaefi** [comment]
@Conan-Kudo are you ok with this state now ? Debian and Ubuntu tests are working fine

**Review by Conan-Kudo — APPROVED** [review]

**Pull request merged**